Excited states

jamie-tay opened the pull request:
Added a new method for the `advance` function that takes multiple states and orthogonalises them. The `orthogonalisation_interval` parameter of the `GramSchmidt` `SpectralStrategy` determines how often the states are orthogonalised.

jamie-tay committed:
Added ability to calculate excited states with Gram-Schmidt orthogonalistion

joachimbrand commented on src/strategies_and_params/spectralstrategy.jl:
```diff
@@ -18,11 +18,12 @@ num_spectral_states(::SpectralStrategy{S}) where {S} = S
-    num_spectral_states > 1 && throw(ArgumentError("num_spectral_states > 1 is currently not supported."))
-    GramSchmidt{num_spectral_states}()
+function GramSchmidt(num_spectral_states = 1, orthogonalisation_interval = 1)
+    return GramSchmidt{num_spectral_states}(orthogonalisation_interval)
```
```suggestion
function GramSchmidt(S = 1; orthogonalization_interval = 1)
    return GramSchmidt{S}(orthogonalization_interval)
```
I suggest making `orthogonalization_interval` a keyword argument. I hope this makes the role of the two arguments a bit less confusing (currently `GramSchmidt(3, 15) == GramSchmidt{3}(15)`. 

It would be great to add a method for `Base.show` that prints it as
`GramSchmidt(3; orthogonalization_interval=15)`

Also, we are trying to stick with American spelling for the code.

jamie-tay replied:
Ok, I'll add a new `Base.show` method.

joachimbrand commented on src/strategies_and_params/spectralstrategy.jl:
```diff
@@ -18,11 +18,12 @@ num_spectral_states(::SpectralStrategy{S}) where {S} = S
 """
-struct GramSchmidt{S} <: SpectralStrategy{S} end
+struct GramSchmidt{S} <: SpectralStrategy{S}
+    orthogonalisation_interval::Int
```
```suggestion
    orthogonalization_interval::Int
```

joachimbrand commented on src/strategies_and_params/spectralstrategy.jl:
```diff
@@ -18,11 +18,12 @@ num_spectral_states(::SpectralStrategy{S}) where {S} = S
     GramSchmidt{S} <: SpectralStrategy{S}
 
 Use the Gram-Schmidt procedure to orthogonalize the excited states. A total of `S` spectral
-states are used in the simulation.
+states are used in the simulation, and they are orthogonalised every `orthogonalisation_interval` steps.
```
We should really document the type constructor, making clear how you create an instance and access the options. Also providing links for context and information how to use it is helpful.

Something like

```
"""
    GramSchmidt(S; orthogonalization_interval = 1) <: SpectralStrategy{S} 
Use the Gram-Schmidt procedure to orthogonalize the excited states. A total of `S` spectral
states are used in the simulation, and they are orthogonalized every 
`orthogonalization_interval` steps.

Use with the keyword argument `spectral_strategy` in [`ProjectorMonteCarloProblem`](@ref).
"""
```

jamie-tay replied:
Ok, I'll add that, thanks.

joachimbrand commented on src/fciqmc.jl:
```diff
@@ -188,4 +188,18 @@ end
 end
-# TODO: add advance! for SpectralState{N} where N > 1
+
+function advance!(algorithm, report, state::ReplicaState, replica::SpectralState{N, NS, GramSchmidt{N}}) where {N, NS}
```
```suggestion
function advance!(algorithm, report, state::ReplicaState, replica::SpectralState{N, <:Any, GramSchmidt{N}}) where {N}
```
I don't think that we need `NS`.

joachimbrand commented on src/fciqmc.jl:
```diff
@@ -188,4 +188,18 @@ end
+    if state.step[] % replica.spectral_strategy.orthogonalisation_interval == 0
+        for i in 1:N
+            for j in 1:i-1
+                add!(replica[i].v, replica[j].v, -1*dot(replica[j].v,replica[i].v)/dot(replica[j].v, replica[j].v), One())
```
```suggestion
                add!(replica[i].v, replica[j].v, -1*dot(replica[j].v,replica[i].v)/dot(replica[j].v, replica[j].v))
```
I think that we don't need the last argument.

Also, please try to keep the length of a line below 92 characters. We are trying to adhere to the 
Invenia Blue style https://github.com/JuliaDiff/BlueStyle

mtsch replied:
I suggest writing it as
```suggestion
                u = replica[i].v
                v = replica[j].v
                add!(u, v, -dot(u, v) / dot(v, v))
```

jamie-tay replied:
Yes that looks better, thank you, I'll try to stick to that style.

joachimbrand commented on src/fciqmc.jl:
```diff
@@ -188,4 +188,18 @@ end
+        proceed &= advance!(algorithm, report, state, replica[i])
+    end
+    return proceed
+end
```
```suggestion
end

```
Please add a newline at the end of the file.

joachimbrand reviewed (changes requested):
Thanks for the nice PR!

I like the idea of orthogonalising the state before the spawning step. It makes life easier for the norm control machinery. It does mean, however, that we cannot rely on the instantaneous states being exactly orthogonal.

It would be great to have this work smoothly with `ProjectorMonteCarloProblem` and `init!` to create an appropriate set of starting vectors. I think currently the number of permissible spectral states is still hardcoded to be one (see line 39 of "pmc_simulation.jl").

It would further be good to make it easy to pass suitable initial state. It would also be super convenient to have suitable initial state generated automatically if no starting vectors (or an insufficient number of them) are provided.

A possible strategy is the following: Say only a single address is available (e.g. from `starting_address(ham)`. Generate further addresses with `offdiagonals` until you have at least `num_spectral_states` of them and then sort them by the value of `diagonal_element`. Set as starting vectors vectors with one of these addresses each. This should work well for diagonally dominated Hamiltonians.

mtsch reviewed (commented):
Very nice!

jamie-tay commented:
> Thanks for the nice PR!
> 
> I like the idea of orthogonalising the state before the spawning step. It makes life easier for the norm control machinery. It does mean, however, that we cannot rely on the instantaneous states being exactly orthogonal.
> 
> It would be great to have this work smoothly with `ProjectorMonteCarloProblem` and `init!` to create an appropriate set of starting vectors. I think currently the number of permissible spectral states is still hardcoded to be one (see line 39 of "pmc_simulation.jl").
> 
> It would further be good to make it easy to pass suitable initial state. It would also be super convenient to have suitable initial state generated automatically if no starting vectors (or an insufficient number of them) are provided.
> 
> A possible strategy is the following: Say only a single address is available (e.g. from `starting_address(ham)`. Generate further addresses with `offdiagonals` until you have at least `num_spectral_states` of them and then sort them by the value of `diagonal_element`. Set as starting vectors vectors with one of these addresses each. This should work well for diagonally dominated Hamiltonians.

Yes it would be good to be able to automatically generate starting vectors, thanks, I'll look into doing that.

jamie-tay committed:
Apply suggestions from code review

Rewrite orthogonalization of vectors, make orthogonalization_interval a keyword argument, fix spelling

Co-authored-by: Joachim Brand <j.brand@massey.ac.nz>
Co-authored-by: mtsch <matijacufar@gmail.com>

mtsch reviewed (changes requested):
Looks great! The next step would be to add some tests. It doesn't have to be anything complicated, just something that makes sure the code works correctly. I suggest something like setting up a small Hamiltonian (maybe `HubbardReal1D(BoseFS(1,1,1,1,1))`), solving it with FCIQMC with the `style` set to `IsDeterministic()`, then comparing the resulting energies and vectors with the result that comes out of `solve`ing an `ExactDiagonalizationProblem`.

If you run into issues getting this to work feel free to ask here or on Zulip!

jamie-tay committed:
Update documentation, add Base.show method for GramSchmidt

jamie-tay committed:
Merge branch 'develop' into feature/excited-states

jamie-tay committed:
Add automatic generation of starting vectors

jamie-tay committed:
Add tests for excited states

jamie-tay committed:
fix imports

mtsch commented on src/pmc_simulation.jl:
```diff
@@ -21,42 +21,57 @@ mutable struct PMCSimulation
             )
         end
+    elseif n_spectral > 1
+        basis = build_basis(ham; cutoff=diagonal_element(ham, starting_address(ham)) + 5)
+        basis = sort!(basis; by=x -> diagonal_element(ham, x))[1:2*n_spectral]
+        trunc = BasisSetRepresentation(ham, basis; filter=Returns(false), sizelim=Inf)
+        vecs = eigvecs(Matrix(trunc))
+        if threading
+            eigs = [PDVec(zip(basis, 10*vecs[:,i]); style, initiator) for i in 1:n_spectral]
+        else
+            if initiator == NonInitiator()
+                eigs = [DVec(zip(basis, 10*vecs[:,i]); style) for i in 1:n_spectral]
+            else
+                eigs = [InitiatorDVec(zip(basis,10*vecs[:,i]);style,initiator) for i in 1:n_spectral]
+            end
+        end
```
I like this idea a lot, but we should be careful. It's possible to setup a Hamiltonian that's huge, but still has all diagonal elements below `diagonal_element(ham, starting_address(ham)) + 5`. Running this on it will not work.

joachimbrand replied:
As discussed, I have the same concern. I would suggest to not use an energy cutoff, but instead implement a new option for `build_basis` where you specify a lower bound, e.g. called `mininum_size`, that allows `build_basis` to stop generating new addresses once the `mininum_size` has been reached. 

This will require modifying the code of `build_basis`, but should be fairly easy to implement.

jamie-tay replied:
I've tried to implement this and it seems to work for me, but one of the tests failed and I'm not understanding what happened - could somebody please take a look?

joachimbrand replied:
The failing test has nothing to do with your (or our) code. A different package which is a dependency of our code (here `REPL.jl`) failed to precompile on the nightly version of Julia. This can happen when things are changed in the bleeding-edge (and unreleased) development version of Julia. Usually these problems will disappear by themselves.

joachimbrand reviewed (commented):
Very nice PR! I've found two issues:
- The new functionality is not yet well documented. I suggest documenting the option of passing a `spectral_strategy` in the docstring of `ProjectorMonteCarloProblem`, and adding a reference to `GramSchmidt` in the docstring of `SpectralStrategy`. I'm not completely sure whether the latter should be exported. The used would not need to use it directly in most cases, but I'd like to properly document the available options (which currently is only `GramSchmidt`.
- I've checked that this works nicely when generating excited state with a single replica. However, combining replicas with excited states throws an error:

```julia-repl
julia> h = HubbardReal1D(near_uniform(BoseFS{10,10}))
HubbardReal1D(fs"|1 1 1 1 1 1 1 1 1 1⟩"; u=1.0, t=1.0)

julia> spectral_strategy = GramSchmidt(3)
GramSchmidt(3; orthogonalization_interval=1)

julia> p = ProjectorMonteCarloProblem(h; spectral_strategy, replica_strategy=AllOverlaps(2), last_step=2_000)
ProjectorMonteCarloProblem with 2 replica(s) and 3 spectral state(s):
  Algorithm: FCIQMC(DoubleLogUpdate{Int64}(1000, 0.08, 0.0016), ConstantTimeStep())
  Hamiltonian: HubbardReal1D(fs"|1 1 1 1 1 1 1 1 1 1⟩"; u=1.0, t=1.0)
  Style: IsDynamicSemistochastic{Float64,ThresholdCompression,DynamicSemistochastic}()
  Initiator: NonInitiator()
  Threading: false
  Simulation Plan: SimulationPlan(starting_step=0, last_step=2000, walltime=Inf)
  Replica Strategy: AllOverlaps{2, 0, Tuple{}, true}(())
  Reporting Strategy: ReportDFAndInfo
  reporting_interval: Int64 1
  info_interval: Int64 100
  io: Base.TTY
  writeinfo: Bool false
  Post Step Strategy: ()
  Spectral Strategy: GramSchmidt(3; orthogonalization_interval=1)
  Maxlength: 2100
  Metadata: OrderedCollections.LittleDict("display_name" => "PMCSimulation")
  Random Seed: 1943348664599472552

julia> sim = solve(p)
ERROR: ArgumentError: Collection has multiple elements, must contain exactly 1 element
Stacktrace:
  [1] only
    @ ./iterators.jl:1527 [inlined]
  [2] #34
    @ ~/git/code/Rimu/src/strategies_and_params/replicastrategy.jl:136 [inlined]
  [3] ntuple
    @ ./ntuple.jl:49 [inlined]
  [4] replica_stats(rs::AllOverlaps{…}, spectral_states::Tuple{…})
    @ Rimu ~/git/code/Rimu/src/strategies_and_params/replicastrategy.jl:136
  [5] step!(sm::Rimu.PMCSimulation)
    @ Rimu ~/git/code/Rimu/src/pmc_simulation.jl:283
  [6] macro expansion
    @ ~/git/code/Rimu/src/pmc_simulation.jl:428 [inlined]
  [7] macro expansion
    @ ~/.julia/packages/ProgressLogging/6KXlp/src/ProgressLogging.jl:328 [inlined]
  [8] 
    @ Rimu ~/git/code/Rimu/src/pmc_simulation.jl:422
  [9] solve!
    @ ~/git/code/Rimu/src/pmc_simulation.jl:337 [inlined]
 [10] solve(args::ProjectorMonteCarloProblem{2, 3})
    @ CommonSolve ~/.julia/packages/CommonSolve/JfpfI/src/CommonSolve.jl:23
 [11] top-level scope
    @ REPL[77]:1
Some type information was truncated. Use `show(err)` to see complete types.

julia> 
```

joachimbrand commented on src/fciqmc.jl:
```diff
@@ -192,4 +192,20 @@ end
+            for j in 1:i-1
+                u = replica[i].v
+                v = replica[j].v
+                add!(u, v, -dot(u, v) / dot(v, v))
```
```suggestion
                add!(u, v, -dot(u, v) / norm(v)^2)
```
`norm` is much faster to evaluate the `dot` for `DVecs`

jamie-tay committed:
Update src/fciqmc.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

jamie-tay committed:
Update documentation, fix imports

jamie-tay committed:
Add minimum_size option for build_basis

joachimbrand commented on src/pmc_simulation.jl:
```diff
@@ -21,42 +21,57 @@ mutable struct PMCSimulation
             )
         end
+    elseif n_spectral > 1
+        basis = build_basis(ham; minimum_size=n_spectral*5)
```
Hmm, I'm not sure whether the factor 5 is reasonable here. It seems a large. The larger this matrix, the better the starting vectors will be, but as soon as these numbers become larger, then `eigvecs` can consume a lot of resources. Maybe 2 is more reasonable?

```suggestion
        basis = build_basis(ham; minimum_size=n_spectral*2)
```
We could also think about allowing the user to set the `minimum_size` by making it a keyword argument in `ProjectorMonteCarloProblem` that is passed through. What do you think?

joachimbrand replied:
Another comment I have here is that the current code ignores the value of `start_at` in this branch. There might be value in allowing to pass an `allowed_address_type`, which can then be passed as a second argument into `build_basis`.

jamie-tay replied:
Ok, I've set the default to a factor of 2, and made `minimum_size` a keyword argument in `ProjectorMonteCarloProblem`. I think that will be useful to allow the option of starting with better approximations to the eigenvectors.

At the moment `build_basis` just uses the `starting_address` of the Hamiltonian, is there a situation where something else should be used?

joachimbrand replied:
```quote
At the moment build_basis just uses the starting_address of the Hamiltonian, is there a situation where something else should be used?
```
I'm thinking of a momentum-conserving momentum-space Hamiltonian (e.g. `HubbardMom1D`) where you could access blocks with different total momentum simply by starting from a different address (which would be chosen to have the total momentum that you want to consider). This could be done by creating a new Hamiltonian with a different starting address, of course, but there might be some value of being able to pass that in later.

The other use case is that you've created a Hamiltonian (possibly strongly correlated) but the `starting_address` is not a very good starting point. What is a good starting point may depend on the value of an interaction strength parameter, for example.

joachimbrand commented on src/pmc_simulation.jl:
```diff
@@ -21,42 +21,57 @@ mutable struct PMCSimulation
         end
+    elseif n_spectral > 1
+        basis = build_basis(ham; minimum_size=n_spectral*5)
+        basis = sort!(basis; by=x -> diagonal_element(ham, x))
```
Why are we sorting the basis here? If we later do a full diagonalisation then there is no point, as the eigenvectors returned by `eigvecs` will be sorted regardless of what the basis looks like.

jamie-tay replied:
Thanks, that's now been removed.

joachimbrand reviewed (changes requested):
Nice work!

See comments below.

To make `AllOverlaps` work with spectral states requires more thoughts about reporting.  I'd be happy to leave that for another PR.

By the way, why were the matrix dimensions swapped (`n_replicas` and `n_spectral`)? Was that inconsistent in the earlier code?

jamie-tay committed:
Add minimum_size keyword argument to ProjectorMonteCarloProblem

jamie-tay commented:
> By the way, why were the matrix dimensions swapped (`n_replicas` and `n_spectral`)? Was that inconsistent in the earlier code?

Yes, the dimensions should now be consistent.

joachimbrand commented on src/projector_monte_carlo_problem.jl:
```diff
@@ -52,6 +52,8 @@ the [`FCIQMC`](@ref) algorithm.
 - `replica_strategy = NoStats(n_replicas)`: Which results to report from replica
   simulations, see [`ReplicaStrategy`](@ref).
+- `spectral_strategy = GramSchmidt()`: The [`SpectralStrategy`](@ref) used for simulations
+  of spectral states.
```
```suggestion
- `n_spectral = 1`: Number of targeted spectral states. Set `n_spectral  > 1` to find excited states.
- `spectral_strategy = GramSchmidt(n_spectral)`: The [`SpectralStrategy`](@ref) used for 
  orthogonalizing spectral states.
```
I would suggest to also implement it as such with introducing a new keyword argument `n_spectral` (similar to how it is done with replicas). This way the user doesn't have to wrap their head around a `spectral_strategy` if they just want a certain number of spectral states.

jamie-tay replied:
Yes this is better, it's now implemented.

joachimbrand commented on src/pmc_simulation.jl:
```diff
@@ -21,42 +21,56 @@ mutable struct PMCSimulation
             )
         end
+    elseif n_spectral > 1
+        basis = build_basis(ham; minimum_size)
```
```suggestion
        basis = build_basis(ham, start_at; minimum_size)
```
As per my earlier comment, we should pass through the information about the starting address.

joachimbrand reviewed (changes requested):
Very nice!

Can we add a test for the `minimum_size` option for `build_basis`?

Using `AllOverlaps` with more than 1 spectral state is not yet implemented and leads to an unhelpful error message. Maybe a check can be added to the constuctor for `ProjectorMonteCarloProblem` to throw an error with a helpful message for this combination?

jamie-tay committed:
Add n_spectral keyword argument to ProjectorMonteCarloProblem

jamie-tay committed:
Pass through start_at to build_basis when generating starting vectors for excited states

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

jamie-tay committed:
Merge branch 'feature/excited-states' of https://github.com/joachimbrand/Rimu.jl into feature/excited-states

jamie-tay committed:
Add error message for AllOverlaps with more than one spectral state

jamie-tay committed:
Resolve conflict

joachimbrand commented on src/projector_monte_carlo_problem.jl:
```diff
@@ -181,6 +184,12 @@ function ProjectorMonteCarloProblem(
 
+    n_spectral = num_spectral_states(spectral_strategy) # spectral_strategy may override n_spectral
+
+    if replica_strategy == AllOverlaps(n_replicas) && n_spectral > 1
```
```suggestion
    if replica_strategy isa AllOverlaps && n_spectral > 1
```
What you wrote also works, but this is a bit more concise and simpler.

joachimbrand reviewed (dismissed):
Looks good now.

mtsch reviewed (dismissed):
Looks good now!

jamie-tay committed:
Update src/projector_monte_carlo_problem.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>
